### ISSUE TITLE
Determine what how we want to report errors to the user when they give bad values

### ISSUE THREAD

**Issue by Insomniak47** [slagroom/Parsimony#8]
We don't seem to have a good set of exceptions that mean some of the things we would want to report back to the user. We should consider either - reading additional docs and then determining which of the built ins is most appropriate for our use case. Came up in #7 

if 

```
  public CommandBuilder(string name)
        {
            if (string.IsNullOrWhiteSpace(name))
                throw new ArgumentException("parameter cannot be null or whitespace", nameof(name));
```

Should be an "ArgumentException" and "ArgumentOutOfRangeException" or a "FormatException". I'm sure this will come up more often so we should maybe just get it sorted sooner rather than later

**Comment by ghost** [comment]
I think ANE/AOOR makes sense in this case. RE the PR comment, I would say there _is_ a set of allowable inputs and, after a null check, any whitespace-only value falls outside of it. The difference I suspect is just in spirit; in that the set is not trivially enumerable. 

I think format exception would make sense if there were a public type that represented the format of string we're accepting.

**Comment by Insomniak47** [comment]
I disagree about AOOR. In that case most common  Argument exceptions could be AOOR. Null -> OOR because the set is [All Non Null Strings]. 

FormatException -> AOOR where set is [All strings matching this pattern]

I think that for argument out of range to be useful there should be either: clearly defined bounds on the set ie [0,len] of an array. A fixed set of strings/enums, or concrete defined bounds on an uninterrupted actual range of values. ie pick a number between 21 and 47". 

I agree that FormatException also isn't a great fit but I think maybe we should just come up with our own in that case.
